Convert "SIG Scalability - Kubernetes The Hard Way" blog to Markdown (#19628)
kubernetes/website: content/en/blog/_posts/2020-03-16-kubernetes-the-hard-way.md

---
layout: blog

> title: SIG Scalability - Learning Kubernetes the Hard Way

Reviewer: ```suggestion
title: Join SIG Scalability and Learn Kubernetes the Hard Way
```

Reviewer: Notes on the first few edits: 

- CTA in title, first paragraph, and closing
- Short, descriptive slug (helps with SEO)

date: 2020-03-16

> slug: sig-scalability-learning-kubernetes-the-hard-way

Reviewer: ```suggestion
slug: join-sig-scalability
```

---

**Authors:** Alex Handy

> 

Reviewer: ```suggestion
Contributing to SIG Scalability is a great way to learn Kubernetes in all its depth and breadth, and the team would love to have you [join as a contributor](https://github.com/kubernetes/community/tree/master/sig-scalability#scalability-special-interest-group). I took a look at the value of learning the hard way and interviewed the current SIG chairs to give you an idea of what contribution feels like.

## The value of Learning The Hard Way:tm:
```

> There is a strain of thinking in the software development community that pushes for the most challenging and rigorous possible method of learning a new language or system. These tend to go by the moniker of "Learn \_\_ the Hard Way." Examples abound: Learn Code the Hard Way, Learn Python the Hard Way, and of course, the one that started the whole movement, Zed Shaw's courses in the topic.

Reviewer: The phrase "strain of thinking" is interesting, but unclear. Is straining a positive or negative? Better to remove the ambiguity and opinion by stating as a data point: 

```suggestion
There is a belief in the software development community that pushes for the most challenging and rigorous possible method of learning a new language or system. These tend to go by the moniker of "Learn \_\_ the Hard Way." Examples abound: [Learn Code the Hard Way](https://shop.learncodethehardway.org/), Learn Python the Hard Way, and many others originating with Zed Shaw's courses in the topic.
```

Author: I believe with 3rd party links, we are limiting them at the current time (do you know if that's true, @jimangel or @zacharysarah?) But if it's okay, I'll commit in that change.

Reviewer 2: > strain of thinking

Is that a misremembering of “train of thought” perhaps?

Author: Ah! Great catch, @sftim. I'm glad you don't take things like that for granite :joy: haha!

Reviewer 2: Alternatively: “strand of thinking” is a less ambiguous rewording that is close to the original.

> While there are folks out there who offer you a "Learn Kubernetes the Hard Way" type experience (most notably Google's Kelsey Hightower), any "Hard Way" project should attempt to cover every aspect of the core topic's principles.

Reviewer: ```suggestion
While there are folks out there who offer you a "Learn Kubernetes the Hard Way" type experience (most notably [Kelsey Hightower's](https://github.com/kelseyhightower/kubernetes-the-hard-way)), any "Hard Way" project should attempt to cover every aspect of the core topic's principles.
```

Reviewer: Suggestion: Link and avoid phrasing a person's individual activity as owned by the company

Therefore, the real way to "Learn Kubernetes the Hard Way," is to join the CNCF and get involved in the project itself. And there is only one SIG that could genuinely offer a full-stack learning experience for Kubernetes: SIG Scalability.

> 

Reviewer: ```suggestion

## What is SIG Scalability anyway?
```

Reviewer: Offering a playful title, feel free to reword in your own tone.

> The team behind SIG Scalability is responsible for detecting and dealing with issues that arise when Kubernetes clusters are working with upwards of a thousand nodes. Said Wojiciech Tyczynski, a staff software engineer at Google and a member of SIG Scalability, the standard size for a test cluster for this SIG is over 5,000 nodes.

Reviewer: ```suggestion
The team behind SIG Scalability is responsible for detecting and dealing with issues that arise when Kubernetes clusters are working with upwards of a thousand nodes. Said [Wojiciech Tyczynski](https://github.com/wojtek-t), a staff software engineer at Google and a member of SIG Scalability, the standard size for a test cluster for this SIG is over 5,000 nodes.
```

And yet, this SIG is not composed of Ph.D.'s in highly scalable systems designs. Many of the folks working with Tyczynski, for example, joined the SIG knowing very little about these types of issues, and often, very little about Kubernetes.

Working on SIG Scalability is like jumping into the deep end of the pool to learn to swim, and the SIG is inherently concerned with the entire Kubernetes project. SIG Scalability focuses on how Kubernetes functions as a whole and at scale. The SIG Scalability team members have an impetus to learn about every system and to understand how all systems interact with one another.

> 

Reviewer: ```suggestion

## A complex and rewarding contributor experience
```

I'm looking to transition this into the next part of the narrative that calls to the reader we want to attract. "Hey, this is hard, and if you like that, then it's for you!" Feel free to adjust.

While that may sound complicated (and it is!), that doesn't mean it's outside the reach of an average developer, tester, or administrator. Google software developer Matt Matejczyk has only been on the team since the beginning of 2019, and he's been a valued member of the team since then, ferreting out bugs.

> "I am new here," said Matejczyk. "I joined the team in January of this year. Before that, I worked on AdWords at Google in New York. Why did I join? I knew some guys there, so that was one of the decisions for me to move. I thought at that time that Kubernetes is a unique, cutting edge technology. I thought it'd be cool to work on that."

Reviewer 2: @vonguard _I realize this is a quote_; however, I wonder it it's feasible to get a similar quote with more inclusive wording?

For example: “people” in place of “guys”.

Author: @mm4tt - would you be okay with us changing that quote accordingly? (I can take care of the markup!)

Reviewer 3: Yeah, totally! I didn't know I was going to be quoted :)

> "I am new here," said Matejczyk. "I joined the team in January of this year. Before that, I worked on AdWords at Google in New York. Why did I join? I knew some guys there, so that was one of the decisions for me to move. I thought at that time that Kubernetes is a unique, cutting edge technology. I thought it'd be cool to work on that."

Reviewer: ```suggestion
"I am new here," said Matejczyk. "I joined the team in January [2020]. Before that, I worked on AdWords at Google in New York. Why did I join? I knew some [people] there, so that was one of the decisions for me to move. I thought at that time that Kubernetes is a unique, cutting edge technology. I thought it'd be cool to work on that."
```

Recommending avoiding time stamps that make it difficult to understand. Also adjusting per @sftim's (great) recommendation.

When Matejczyk spoke to what he had worked on during those 8 months, he answered, "Mostly detecting and debugging regressions. This gets really interesting usually, as it takes a lot of time to debug everything from end-to-end, but when you figure out everything, it can be rewarding. What is interesting is that it looks like everything boils down to memory allocation. That is one piece of code, and part of the same binary can affect other code only because it is allocating too much memory."

Tyczynski said that "It's not only debugging regressions, but it's also debugging and finding bottlenecks. In general, those can be regressions, but those can be things we can improve. The other significant area is extending what we want to guarantee to users. Extending SLA and SLO coverage of the system so users can rely on what they can expect from the system in terms of performance and scalability. Matt is doing much work in extending our tests to be more representative and cover more Kubernetes concepts."

> 

Reviewer: ```suggestion

## Give SIG Scalability a try

```

> The SIG Scalability team is always in need of new members, and if you're the sort of developer or tester who loves taking on new complex challenges, and perhaps loves learning things the hard way, consider joining this SIG. As the team points out, adding Kubernetes expertise to your resume is never a bad idea, and this is the one SIG where you can learn it all from top to bottom.

Reviewer: ```suggestion
The SIG Scalability team is always in need of new members, and if you're the sort of developer or tester who loves taking on new complex challenges, and perhaps loves learning things the hard way, consider joining this SIG. As the team points out, adding Kubernetes expertise to your resume is never a bad idea, and this is the one SIG where you can learn it all from top to bottom.

See [the SIG's documentation](https://github.com/kubernetes/community/tree/master/sig-scalability#scalability-special-interest-group) to learn about upcoming meetings, its charter, and more. You can also join the [#sig-scalability Slack channel](https://kubernetes.slack.com/archives/C09QZTRH7) to see what it's like. We hope to see you join in to take advantage of this great opportunity to learn Kubernetes and contribute back at the same time.
```

A bit of a soft pitch at the end but better than nothing 😉

---
layout: blog
title: Join SIG Scalability and Learn Kubernetes the Hard Way
date: 2020-03-16
slug: join-sig-scalability
---

**Authors:** Alex Handy

> Contributing to SIG Scalability is a great way to learn Kubernetes in all its depth and breadth, and the team would love to have you [join as a contributor](https://github.com/kubernetes/community/tree/master/sig-scalability#scalability-special-interest-group). I took a look at the value of learning the hard way and interviewed the current SIG chairs to give you an idea of what contribution feels like.

Reviewer 4: this is coming through as on the same line as alex

Author: Ah! Thank you - a great catch!

And yet, this SIG is not composed of Ph.D.'s in highly scalable systems designs. Many of the folks working with Tyczynski, for example, joined the SIG knowing very little about these types of issues, and often, very little about Kubernetes.

Working on SIG Scalability is like jumping into the deep end of the pool to learn to swim, and the SIG is inherently concerned with the entire Kubernetes project. SIG Scalability focuses on how Kubernetes functions as a whole and at scale. The SIG Scalability team members have an impetus to learn about every system and to understand how all systems interact with one another.

## A complex and rewarding contributor experience

While that may sound complicated (and it is!), that doesn't mean it's outside the reach of an average developer, tester, or administrator. Google software developer Matt Matejczyk has only been on the team since the beginning of 2019, and he's been a valued member of the team since then, ferreting out bugs.

> "I am new here," said Matejczyk. "I joined the team in January [2020]. Before that, I worked on AdWords at Google in New York. Why did I join? I knew some [people] there, so that was one of the decisions for me to move. I thought at that time that Kubernetes is a unique, cutting edge technology. I thought it'd be cool to work on that."

Reviewer 2: I'm not clear on whether @mm4tt joined at the start of 2019 or the start of 2020. There's a big difference in the amount of onboarding time this implies so I'm calling it out.

(not hard to fix after the blog post goes out IMO)

Author: Could I get your help in clarifying, @mm4tt?

Reviewer: I see Matt is on sick leave according to his GitHub profile, so I did a quick search. I see reference to him contributing to SIG Scalability in notes from last year, which confirms 2019 is a better date to put. My hope was it was more recent since it would reinforce the narrative that people can quickly learn to contribute to the SIG. Alas, accuracy is more important 😄 

I say we switch that date up and moving forward @onlydole.

> "I am new here," said Matejczyk. "I joined the team in January [2020]. Before that, I worked on AdWords at Google in New York. Why did I join? I knew some [people] there, so that was one of the decisions for me to move. I thought at that time that Kubernetes is a unique, cutting edge technology. I thought it'd be cool to work on that."

Reviewer: ```suggestion
"I am new here," said Matejczyk. "I joined the team in January [2019]. Before that, I worked on AdWords at Google in New York. Why did I join? I knew some [people] there, so that was one of the decisions for me to move. I thought at that time that Kubernetes is a unique, cutting edge technology. I thought it'd be cool to work on that."
```

Thinking more about this.. I do like that it only took 1.5 year to be a Team Lead and (arguably) an expert on the subject. That's an exciting subtext.

Reviewer: @onlydole confirmed 2019 👍

> "I am new here," said Matejczyk. "I joined the team in January [2020]. Before that, I worked on AdWords at Google in New York. Why did I join? I knew some [people] there, so that was one of the decisions for me to move. I thought at that time that Kubernetes is a unique, cutting edge technology. I thought it'd be cool to work on that."

Reviewer 2: Combining suggestions including feedback in https://github.com/kubernetes/website/pull/19628#discussion_r392931287
```suggestion
"I am new here," said Matejczyk. "I joined the team in January [2019]. Before that, I worked on AdWords at Google in New York. Why did I join? I knew some people there, so that was one of the decisions for me to move. I thought at that time that Kubernetes is a unique, cutting edge technology. I thought it'd be cool to work on that."
```

Matejczyk was correct about the coolness. "It's cool," he said. "So actually, ramping up on scalability is not easy. There are many things you need to understand. You need to understand Kubernetes very well. It can use every part of Kubernetes. I am still ramping up after these 8 months. I think it took me maybe 3 months to get up to decent speed."

> When Matejczyk spoke to what he had worked on during those 8 months, he answered, "Mostly detecting and debugging regressions. This gets really interesting usually, as it takes a lot of time to debug everything from end-to-end, but when you figure out everything, it can be rewarding. What is interesting is that it looks like everything boils down to memory allocation. That is one piece of code, and part of the same binary can affect other code only because it is allocating too much memory."

Reviewer 3: I would rephrase the last two sentences. I don't remember what I exactly said, but I'm quite sure I was giving an example and refreezing to a particular bottleneck I was working on during that time. In the current form, a reader may get an impressions that every k8s performance regressions / bottleneck boils down to memory allocation, which is not true.

What about something like this?

> An interesting example is the regression I have been working on recently. We noticed overall slowness of Kubernetes control plane in certain scenarios and we couldn't attribute it to any particular component. In the end, we realized that everything boiled down to the memory allocation on the golang level. It was very counterintuitive to have two completely separate pieces of code (running as a part of the same binary) affecting performance of each other only because one of them was allocating memory too fast. But connecting all the dots and getting to the bottom of regression like this gives a lot of satisfaction.

It's just an example, feel free to rephrase appropriately.
